docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,995 @@
+        uint32_t flags,
+        const oe_claim_t* custom_claims,
+        size_t custom_claims_length,
+        const void* opt_params,
+        size_t opt_params_size,
+        uint8_t** evidence_buffer,
+        size_t* evidence_buffer_size,
+        uint8_t** endorsements_buffer,
+        size_t* endorsements_buffer_size);
+
+    /**
+     * Frees the generated attestation evidence and endorsements.
+     *
+     * @param[in] plugin_context A pointer to the attester plugin struct.
+     * @param[in] evidence_buffer A pointer to the evidence buffer.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*free_evidence)(
+        oe_attester_plugin_t* plugin_context,
+        uint8_t* evidence_buffer);
+
+    /**
+     * Frees the generated attestation endorsements.
+     *
+     * @param[in] plugin_context A pointer to the attester plugin struct.
+     * @param[in] endorsements_buffer A pointer to the endorsements buffer.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*free_endorsements)(
+        oe_attester_plugin_t* plugin_context,
+        uint8_t* endorsements_buffer);
+};
+```
+
+Here is the rationale for each element in the plugin struct:
+- `format_id`
+  - Each plugin needs a unique identifier to distinguish itself.
+- `on_register` and `on_unregister`
+  - A plugin might require some setup or teardown when it is registered or
+    unregistered, so these functions are required. Furthermore, a plugin
+    might require configuration, which is why there is a `config_data`
+    parameter. The configuration data can be plugin specific, so no format is
+    specified in this proposal.
+- `get_evidence`, `free_evidence`, and `free_endorsements`
+  - Producing evidence and endorsements is necessary for attestation.
+  - `flags` field to determine local vs. remote attestation.
+  - There is a `custom_claims` parameter because many attestation protocols
+    require the enclave to sign some data from a relying party. For example,
+    many protocols follow the "challenge response" architecture, which requires
+    the enclave to sign a nonce from the relying party.
+  - There is an `opt_params` field because some plugins might require plugin
+    specific input. For example, the SGX local attestation needs the
+    other enclave's target info struct.
+  - There is an `endorsements` parameter to return the endorsements that are
+    coupled with the evidence to ensure that the evidence and endorsements are
+    in sync.
+
+### Verfier Plugin API (Supported by host and enclave side)
+
+The plugin API is very similar to the attester plugin. The only difference
+is that it implements a `verify_evidence` function instead of `get_evidence`.
+
+
+```C
+/**
+ * Claims struct used for claims parameters for the plugin.
+ */
+struct oe_claim_t
+{
+    char* name;
+    uint8_t* value;
+    size_t value_size;
+};
+
+/**
+ * Supported policies for validation. Only time is supported for now.
+ */
+enum oe_policy_type_t
```

Comment:
Why is this not a generic attestation policy that can target any combination of claims?